Core/AVIDump.cpp

```diff
@@ -0,0 +1,360 @@
+			s_stream->codec->time_base, s_stream->time_base);
+	}
+	if (s_stream->codec->coded_frame->key_frame)
+		pkt->flags |= AV_PKT_FLAG_KEY;
+	pkt->stream_index = s_stream->index;
+}
+
+static const u8 *ConvertBufferTo888RGB(const GPUDebugBuffer &buf, u8 *&temp, u32 &w, u32 &h) {
+	// The temp buffer will be freed by the caller if set, and can be the return value.
+	temp = nullptr;
+
+	w = std::min(w, buf.GetStride());
+	h = std::min(h, buf.GetHeight());
+
+	const u8 *buffer = buf.GetData();
+	if (buf.GetFlipped() && buf.GetFormat() == GPU_DBG_FORMAT_888_RGB) {
+		// Silly OpenGL reads upside down, we flip to another buffer for simplicity.
+		temp = new u8[3 * w * h];
+		for (u32 y = 0; y < h; y++) {
+			memcpy(temp + y * w * 3, buffer + (buf.GetHeight() - y - 1) * buf.GetStride() * 3, w * 3);
+		}
+		buffer = temp;
+	}
+	else if (buf.GetFormat() != GPU_DBG_FORMAT_888_RGB) {
+		// Let's boil it down to how we need to interpret the bits.
+		int baseFmt = buf.GetFormat() & ~(GPU_DBG_FORMAT_REVERSE_FLAG | GPU_DBG_FORMAT_BRSWAP_FLAG);
+		bool rev = (buf.GetFormat() & GPU_DBG_FORMAT_REVERSE_FLAG) != 0;
+		bool brswap = (buf.GetFormat() & GPU_DBG_FORMAT_BRSWAP_FLAG) != 0;
+		bool flip = buf.GetFlipped();
+
+		temp = new u8[3 * w * h];
+
+		// This is pretty inefficient.
+		const u16 *buf16 = (const u16 *)buffer;
+		const u32 *buf32 = (const u32 *)buffer;
+		for (u32 y = 0; y < h; y++) {
+			for (u32 x = 0; x < w; x++) {
+				u8 *dst;
+				if (flip) {
+					dst = &temp[(h - y - 1) * w * 3 + x * 3];
+				}
+				else {
+					dst = &temp[y * w * 3 + x * 3];
+				}
+
+				u8 &r = brswap ? dst[2] : dst[0];
+				u8 &g = dst[1];
+				u8 &b = brswap ? dst[0] : dst[2];
+
+				u32 src;
+				switch (baseFmt) {
+				case GPU_DBG_FORMAT_565:
+					src = buf16[y * buf.GetStride() + x];
+					if (rev) {
+						src = bswap16(src);
+					}
+					r = Convert5To8((src >> 0) & 0x1F);
+					g = Convert6To8((src >> 5) & 0x3F);
+					b = Convert5To8((src >> 11) & 0x1F);
+					break;
+				case GPU_DBG_FORMAT_5551:
+					src = buf16[y * buf.GetStride() + x];
+					if (rev) {
+						src = bswap16(src);
+					}
+					r = Convert5To8((src >> 0) & 0x1F);
+					g = Convert5To8((src >> 5) & 0x1F);
+					b = Convert5To8((src >> 10) & 0x1F);
+					break;
+				case GPU_DBG_FORMAT_4444:
+					src = buf16[y * buf.GetStride() + x];
+					if (rev) {
+						src = bswap16(src);
+					}
+					r = Convert4To8((src >> 0) & 0xF);
+					g = Convert4To8((src >> 4) & 0xF);
+					b = Convert4To8((src >> 8) & 0xF);
+					break;
+				case GPU_DBG_FORMAT_8888:
+					src = buf32[y * buf.GetStride() + x];
+					if (rev) {
+						src = bswap32(src);
+					}
+					r = (src >> 0) & 0xFF;
+					g = (src >> 8) & 0xFF;
+					b = (src >> 16) & 0xFF;
+					break;
+				default:
+					ERROR_LOG(COMMON, "Unsupported framebuffer format for screenshot: %d", buf.GetFormat());
+					return nullptr;
+				}
+			}
+		}
+		buffer = temp;
+	}
+
+	return buffer;
+}
+
+void AVIDump::AddFrame()
+{
+	GPUDebugBuffer buf;
```

Comment:
Recommend placing this as a static in the file.  It'll avoid reallocating the buffer each frame.

-[Unknown]